torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -75,9 +75,54 @@ class ConditionalFromPredicateModifier : public kir::IrVisitor {
       setWritePredicate(expr, conditional);
     }
 
+    // Note: [Predicate Inversion for CpAsync]
+    // Today for vectorized support the pattern is:
+    // Initialize buffer -> predicated load
+    // For memcpy async:
+    //    If we initialized and then loaded (without sync) it would be undefined
+    //    behavior.
+    // Initialize only the "virtual out of boundary" accesses.
+    //  Memory allocated, but outside the virtual tensor space.
+    //  Virtual tensor space today is effectively what would be allocated in
+    //  global memory. Then only copy the "within bound" accesses.
+    // This is a WAR today based on how our system is set up.
+    //    We would want to have a separate concept of SMEM space from Virtual or
+    //    GMEM space, so that we know we're only working with the allocated
+    //    SMEM.
+    //  If we hit outside the allocated SMEM bad things happen.
+    // Today asserting in predicate removal making sure that the virtual and
+    // SMEM boundaries line up based on the IterDomains.
+    //
+    // TODO: in a follow up we need to extend the predicate
+    //  infrastructure to generate predicate for both gmem
+    //  and smem, and the predicate removal will need to
+    //  be extended as well for the perf critical regions.
+    if (isPredicatedInitForCpAsync(expr)) {
+      invertPredicateForGmemToSharedMemInitialize(expr);
+    }
+
     kir::IrVisitor::handle(expr);
   }
 
+  // Invert the predicate of given expr.
```

Comment:
The limitation here seems to be that we're predicating anything outside the "valid" regions as being the initialization regions. This is fine as long as we never go outside the allocated regime. However what we'd likely want is predicates that are true only outside the "valid region" and within the "allocated region". Is this understanding correct?

Reply:
Yes that's right. For best reliability in all cases we want to cover we'd need to find precisely the region that's allocated but invalid. But as mentioned above, this is mostly for functionality coverage, and in most perf critical cases we should be able to find alternatives to doing this, which will be in follow ups on cp.async.